File: perl/ch01/p1_2.pl
#!/usr/bin/perl

# Page 22, problem 1.2.

use strict;
use warnings;

sub find_bounds {
    my @arr = @_;
    my ($left, $right) = ($#arr, 0);
    my ($max_seen, $min_seen);

    # Get the right bound:
    $max_seen = $arr[0];
    for my $i (1 .. $#arr) {
        $max_seen = ($max_seen < $arr[$i]) ? $arr[$i] : $max_seen;
        if ($arr[$i] < $max_seen) {
            $right = $i
        }
    }
    # Get the left bound:
    $min_seen = $arr[-1];
    for (my $i = ($#arr - 1); $i >= 0; $i--) {
        $min_seen = ($min_seen > $arr[$i]) ? $arr[$i] : $min_seen;
        if ($arr[$i] > $min_seen) {
            $left = $i;
        }
    }

    return ($left, $right);
}

my @bounds;

if (@ARGV) {
    @bounds = find_bounds(@ARGV);
} else {
    @bounds = find_bounds(qw(3 7 5 6 9));
}

print "(@bounds)\n";

File: java/ch01/p1_2.java
/*
    Page 22, problem 1.2.
 */

import java.util.Arrays;

public class p1_2 {
    public p1_2() {}

    private static int[] find_bounds(int[] arr) {
        int left = arr.length, right = 0;
        int max_seen = Integer.MIN_VALUE, min_seen = Integer.MAX_VALUE;

        // Get the right bound:
        for (int i = 0; i < arr.length; i++) {
            max_seen = Integer.max(max_seen, arr[i]);
            if (arr[i] < max_seen) {
                right = i;
            }
        }
        // Get the left bound:
        for (int i = arr.length - 1; i >= 0; i--) {
            min_seen = Integer.min(min_seen, arr[i]);
            if (arr[i] > min_seen) {
                left = i;
            }
        }

        return new int[]{left, right};
    }

    public static void main(String[] args) {
        int[] list;

        if (args.length > 0) {
            list = new int[args.length];
            for (int i = 0; i < args.length; i++) {
                list[i] = Integer.parseInt(args[i]);
            }
        } else {
            list = new int[]{3, 7, 5, 6, 9};
        }

        System.out.println(Arrays.toString(find_bounds(list)));
    }
}
